{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SUB-BENCHMARK3: Comparing jDR methods on single-cell datasets\n",
    "\n",
    "The performances of the 9 jDR methods are here compared based on their ability to cluster cells based on their cancer cell line of origine. The clustering is performed jointly considering scRNA-seq and scATAC-seq data. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data preprocessing\n",
    "First the data are read in their original format and adapted to be read as input of our run_factorization function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data and processing\n",
    "\n",
    "# Load RNA-seq data\n",
    "exp <- readRDS(\"../data/single-cell/CellLines_RNAseqCounts.RDS\", refhook = NULL) #ENS for genes and counts\n",
    "# Apply log2 on RNA-seq data\n",
    "exp <- log2(exp+1)\n",
    "# Load ATAC-seq data\n",
    "atac_counts<-readRDS(\"../data/single-cell/CellLines_ATACseqCounts.RDS\", refhook = NULL) # peaks counts\n",
    "# Load metadata\n",
    "metadata<-readRDS(\"../data/single-cell/CellLines_metadata.RDS\", refhook = NULL)\n",
    "# Rename columns from metadata\n",
    "colnames(atac_counts) <- metadata[,1]\n",
    "\n",
    "# Export RNA-seq data as tab-separated table\n",
    "write.table(exp, \"../data/single-cell/CellLines_RNAseqCounts.txt\", \n",
    "            sep=\"\\t\", col.names=TRUE, row.names=TRUE)\n",
    "# Add a name (\"probe\") to the first column\n",
    "system(\"sed -i '1s/^/probe\\t/' ../data/single-cell/CellLines_RNAseqCounts.txt\")\n",
    "# Export ATAC-seq data as tab-separated table\n",
    "write.table(atac_counts, \"../data/single-cell/CellLines_ATACseqCounts.txt\", \n",
    "            sep=\"\\t\", col.names=TRUE, row.names=TRUE)\n",
    "# Add a name (\"probe\") to the first column\n",
    "system(\"sed -i '1s/^/probe\\t/' ../data/single-cell/CellLines_ATACseqCounts.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running comparison\n",
    "Two factor are then detected for each jDR method and the distribution of the cells with respect of Factor1 and Factor2 is plotted as a scatter plot. The obtained plots are available in the Results folder. The capability of the different jDR methods to cluster the cells accoridng to their cell line of origin is finally evaluated through the C-index, whose value is reported in the Results folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: MASS\n",
      "Loading required package: NMF\n",
      "Loading required package: pkgmaker\n",
      "Loading required package: registry\n",
      "\n",
      "Attaching package: ‘pkgmaker’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    isFALSE\n",
      "\n",
      "Loading required package: rngtools\n",
      "Loading required package: cluster\n",
      "NMF - BioConductor layer [OK] | Shared memory capabilities [NO: bigmemory] | Cores 7/8\n",
      "  To enable shared memory capabilities, try: install.extras('\n",
      "NMF\n",
      "')\n",
      "Loading required package: mclust\n",
      "Package 'mclust' version 5.4.5\n",
      "Type 'citation(\"mclust\")' for citing this R package in publications.\n",
      "Loading required package: InterSIM\n",
      "Loading required package: tools\n",
      "Loading required package: ade4\n",
      "\n",
      "Attaching package: ‘ade4’\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    score\n",
      "\n",
      "\n",
      "Attaching package: ‘GPArotation’\n",
      "\n",
      "The following object is masked from ‘package:NMF’:\n",
      "\n",
      "    entropy\n",
      "\n",
      "\n",
      "Attaching package: ‘MOFAtools’\n",
      "\n",
      "The following objects are masked from ‘package:NMF’:\n",
      "\n",
      "    featureNames, featureNames<-, predict, sampleNames, sampleNames<-\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    featureNames, featureNames<-, sampleNames, sampleNames<-\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    predict\n",
      "\n",
      "Loading required package: JADE\n",
      "Loading required package: lattice\n",
      "Loading required package: caTools\n",
      "Loading required package: gdata\n",
      "gdata: read.xls support for 'XLS' (Excel 97-2004) files ENABLED.\n",
      "\n",
      "gdata: Unable to load perl libaries needed by read.xls()\n",
      "gdata: to support 'XLSX' (Excel 2007+) files.\n",
      "\n",
      "gdata: Run the function 'installXLSXsupport()'\n",
      "gdata: to automatically download and install the perl\n",
      "gdata: libaries needed to support Excel XLS and XLSX formats.\n",
      "\n",
      "Attaching package: ‘gdata’\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    nobs\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    object.size\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    startsWith\n",
      "\n",
      "Loading required package: gtools\n",
      "\n",
      "Attaching package: ‘gtools’\n",
      "\n",
      "The following object is masked from ‘package:InterSIM’:\n",
      "\n",
      "    logit\n",
      "\n",
      "Loading required package: gplots\n",
      "\n",
      "Attaching package: ‘gplots’\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    lowess\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "Saving 6.67 x 6.67 in image\n",
      "Saving 6.67 x 6.67 in image\n",
      "Saving 6.67 x 6.67 in image\n",
      "Saving 6.67 x 6.67 in image\n",
      "Saving 6.67 x 6.67 in image\n",
      "Saving 6.67 x 6.67 in image\n",
      "Saving 6.67 x 6.67 in image\n"
     ]
    }
   ],
   "source": [
    "library(\"ggplot2\")\n",
    "library(\"clusterCrit\")\n",
    "source(\"runfactorization.R\")\n",
    "\n",
    "# Parameters for the plots\n",
    "dot_size <- 1.5\n",
    "dot_alpha <- 1.0\n",
    "xlabel <- \"Factor 1\"\n",
    "ylabel <- \"Factor 2\"\n",
    "\n",
    "# Load annotations from the metadata\n",
    "sample_annot <- metadata[, c(\"sample.rna\", \"celltype\")]\n",
    "\n",
    "# Folder for results\n",
    "results_folder <- \"../results_single_cell/\"\n",
    "# Create output folder\n",
    "dir.create(results_folder, showWarnings = FALSE)\n",
    "\n",
    "# Run factorization methods\n",
    "out <- runfactorization(\"../data/single-cell/\",\n",
    "                        c(\"CellLines_RNAseqCounts.txt\", \"CellLines_ATACseqCounts.txt\"),\n",
    "                        2, \n",
    "                        sep=\"\\t\", \n",
    "                        filtering=\"stringent\")\n",
    "\n",
    "c_index <- numeric(0)\n",
    "\n",
    "# For each factorization method\n",
    "for(i in 1:length(out$factorizations)){\n",
    "    \n",
    "    # Get factorization result\n",
    "    factors <- out$factorizations[[i]][[1]]\n",
    "\n",
    "    # Delete NAs\n",
    "    factors <- factors[!is.na(factors[,1]) & !is.na(factors[,2]), ]\n",
    "    sample_annot <- sample_annot[!is.na(sample_annot[,1]) & !is.na(sample_annot[,2]), ]\n",
    "\n",
    "    # Data to be plotted\n",
    "    df <- data.frame(x =  factors[,1], y = factors[,2], color_by = sample_annot[,2])\n",
    "    # Plot results\n",
    "    p <- ggplot(df, aes_string(x = \"x\", y = \"y\")) + \n",
    "       geom_point(aes_string(color = \"color_by\"), size=dot_size, alpha=dot_alpha) + \n",
    "       xlab(xlabel) + ylab(ylabel) +\n",
    "       # scale_shape_manual(values=c(19,1,2:18)[seq_along(unique(shape_by))]) +\n",
    "       theme(plot.margin = margin(20, 20, 10, 10), \n",
    "             axis.text = element_text(size = rel(1), color = \"black\"), \n",
    "             axis.title = element_text(size = 16), \n",
    "             axis.title.y = element_text(size = rel(1.1), margin = margin(0, 10, 0, 0)), \n",
    "             axis.title.x = element_text(size = rel(1.1), margin = margin(10, 0, 0, 0)), \n",
    "             axis.line = element_line(color = \"black\", size = 0.5), \n",
    "             axis.ticks = element_line(color = \"black\", size = 0.5),\n",
    "             panel.border = element_blank(), \n",
    "             panel.grid.major = element_blank(),\n",
    "             panel.grid.minor = element_blank(), \n",
    "             panel.background = element_blank(),\n",
    "             legend.key = element_rect(fill = \"white\"),\n",
    "             legend.text = element_text(size = 16),\n",
    "             legend.title = element_text(size =16)\n",
    "       )\n",
    "    p + scale_color_manual(values=c(\"#0072B2\", \"#D55E00\", \"#CC79A7\"))\n",
    "    # Export plot as JPEG image\n",
    "    ggsave(paste0(results_folder, \"plot_\",out$method[i],\".jpg\"))\n",
    "\n",
    "    # Encode cell type annotations by numeric codes\n",
    "    ann <- factor(sample_annot[,2], levels=c(\"HCT\", \"Hela\", \"K562\"))\n",
    "    ann <- as.integer(ann)\n",
    "    # Compare factors and annotations\n",
    "    c_index <- c(c_index, intCriteria(factors, as.integer(ann), crit=c(\"C_index\"))$c_index)\n",
    "\n",
    "}\n",
    "\n",
    "# Build output table\n",
    "report_cindex <- data.frame(method=out$method, cindex=c_index)\n",
    "\n",
    "# Export results as one tab-separated table\n",
    "write.table(report_cindex, file = paste0(results_folder, \"singlecell_cindex.txt\"), \n",
    "            sep=\"\\t\", col.names=FALSE, row.names=FALSE, quote=FALSE)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
